grails-app/services/com/secured/api/DocumentShipmentService.groovy

```diff
@@ -0,0 +1,89 @@
+            {
+                boolean contains = document.products?.contains(products)
+                DevCycleLogger.log("found product with code ${it.product_code}, belonging to company with id ${company.companyId}")
+                products = super.ship(it)
+                if(!products)
+                {
+                    DevCycleLogger.log("couldn't update product with code ${it.product_code}, belonging to company with id ${company.companyId}, rejecting")
+                    response.rejectProduct(it)
+                    response.reportInvalidInput()
+                    continue
+                }
+                if(!contains)
+                {
+                    document.products.add(products) // necessary, hasMany in document is not defined
+                }
+                continue
+            }
+            DevCycleLogger.log("couldn't find product with code ${it.product_code}, belonging to company with id ${company.companyId}, rejecting")
+            response.rejectProduct(it)
+            response.reportInvalidInput()
+        }
+        DevCycleLogger.log("exiting shipProducts()")
+        return response.getAsMap()
+    }
+
+
+    Map ship(ShipmentDocumentCommand cmd)
+    {
+        DevCycleLogger.log("ship() called")
+        Company company = Company.findWhere(token: cmd.companyToken)
+        if(!company)
+        {
+            DevCycleLogger.log("company with token ${cmd.companyToken} not found, reporting authorization failure, exiting ship()")
+            Response response = new Response()
+            response.rejectCompanyToken()
+            return response.getAsMap()
+        }
+        if(!cmd.validate())
+        {
+            DevCycleLogger.log("command object not validated, reporting invalid input, exiting ship()")
+            Response response = new Response()
+            response.reportInvalidInput()
+            cmd.products.each {
+                response.rejectProduct(it)
+            }
+            return response.getAsMap()
+        }
+        Document document = createShipmentDocumentMock(cmd)
+        Map response = shipProducts(document, company, cmd.products)
+        if(!document.products.isEmpty())
+        {
+            if(!document.validate())
+            {
+                document.errors.fieldErrors.each {
+                    DevCycleLogger.log("couldn't validate value ${it.rejectedValue} for field ${it.field}")
+                }
+                DevCycleLogger.log("document with number ${document.documentNumber} not validated, exiting accept()")
+                return response
+            }
+            document.save(true)
+            DevCycleLogger.log("saving document with number ${document.documentNumber}, exiting accept()")
+            return response
+        }
+        DevCycleLogger.log("document with number ${document.documentNumber} not saved(empty products list), exiting accept()")
+        return response
```

Comment:
please implement a new  function for this part and also use response class as we discussed